src/app/reducers/manuscript-editor.reducer.ts

```diff
@@ -1,70 +1,27 @@
-import * as manuscriptActions from '../actions/manuscript.actions';
-import { cloneManuscript, ManuscriptHistory } from '../utils/state.utils';
-import { Transaction } from 'prosemirror-state';
-import { redoChange, undoChange, updateManuscriptState } from '../utils/history.utils';
-import { KeywordAddPayload, KeywordDeletePayload, KeywordUpdatePayload } from '../actions/manuscript.actions';
-import { Action } from '../utils/action.utils';
+import * as manuscriptEditorAction from '../actions/manuscript-editor.actions';
+import { ManuscriptEditorState } from '../store';
 
-export function manuscriptEditorReducer(
-  state: ManuscriptHistory | undefined,
-  action: manuscriptActions.ActionType
-): ManuscriptHistory {
-  if (!state) {
-    return state;
-  }
+const initialState: ManuscriptEditorState = {
+  focusedManuscriptPath: undefined
+};
 
+export function manuscriptEditorReducer(
+  state: ManuscriptEditorState = initialState,
+  action: manuscriptEditorAction.ActionType
+): ManuscriptEditorState {
   switch (action.type) {
-    case manuscriptActions.updateTitleAction.type:
-      return updateManuscriptState(state, 'title', action.payload as Transaction);
-    case manuscriptActions.updateAbstractAction.type:
-      return updateManuscriptState(state, 'abstract', action.payload as Transaction);
-
-    case manuscriptActions.updateKeywordsAction.type:
-      const updatePayload = action.payload as KeywordUpdatePayload;
-      const keywordPath = ['keywords', updatePayload.keywordGroup, updatePayload.index].join('.');
-      return updateManuscriptState(state, keywordPath, updatePayload.change);
-
-    case manuscriptActions.deleteKeywordAction.type:
-      return handleDeleteKeywordAction(state, action as Action<KeywordDeletePayload>);
-
-    case manuscriptActions.addNewKeywordAction.type:
-      return handleAddNewKeywordAction(state, action as Action<KeywordAddPayload>);
-
-    case manuscriptActions.undoAction.type:
-      return undoChange(state);
-    case manuscriptActions.redoAction.type:
-      return redoChange(state);
-    default:
-      return state;
+    case manuscriptEditorAction.setFocusAction.type:
```

Comment:
Do we need to worry about the order that these events could be dispatched in? E.g. if I unfocus a control by focusing another, are the events guaranteed to arrive in the order removeFocus -> setFocus? If not, then it might be an idea to have the removeFocus action have an argument for the name, and only clear the focusedManuscriptPath variable if it's current value matches the name in the payload to prevent it from unfocusing another component?

Reply:
I tested it with our current code. And it works fairly well. It consistently fires blur before new focus. I will test this with other browsers too.